docs/architecture/adr-001-umee-module.md

```diff
@@ -0,0 +1,90 @@
+in a one-to-one ratio, i.e. one ATOM mints one meToken derivative.
+
+Validators on the Umee network will take these deposited ATOM tokens and delegate
+them to a set of governance-controlled validators on the ATOM source chain,
+e.g. the Cosmos Hub.
+
+A user that sent ATOM tokens to the Umee network can then take their derivative
+meTokens and send them to Ethereum via a bridge where a synthetic ERC-20 version
+of the meTokens are minted. Once sent, the meTokens are locked in Umee and the
+user can then freely trade and operate within Ethereum's DeFi ecosystem with the
+synthetic ERC-20 meTokens.
+
+## Decision
+
+To support users sending ATOM tokens to the Umee network via IBC, we will create
+a `x/umee` module that contains a dedicated `ModuleAccount` that represents the
+ATOM pool. In addition, we will implement a [custom IBC module](https://github.com/cosmos/ibc-go/blob/v1.0.0-alpha1/docs/custom.md).
+The custom IBC module will handle receiving custom packets on a custom Umee port.
+The custom module will be nearly identical to the ICS-20 module, except that it
+will perform the following:
+
+- Automatically send source ATOM tokens to the dedicated `ModuleAccount` pool
+  address instead of a user supplied address.
+- Mark the transfer's amount in a pending queue, `PendingDelegationQueue`, to be
+  later used for delegation on the source chain.
+- Track the sender and amount in the module's state.
+- Mint derivative meTokens and send them to the source sender's Umee account.
+
+At the beginning of a staking epoch, E<sub>1</sub> in `BeginBlock`, the `x/umee`
+module will perform the following:
+
+- Move the funds from the `PendingDelegationQueue` into another queue,
+  `PrevPendingDelegationQueue`. This queue is locked as funds cannot be sent to
+  it. In addition, these are the funds that will be delegated to the source chain
+  during the current staking epoch.
+- Clear out `PendingDelegationQueue`.
+
+The validators in the current staking epoch, E<sub>1</sub>, will then construct,
+sign, and broadcast multisig transactions to the Umee network. Once enough
+transactions are received by the `x/umee` module, specifically 2/3 of total
+staking weight, the `x/umee` module will store the constructed multisig transaction
+and emit an event signalling a delegation transaction can be made to the source
+chain.
+
+A separate relayer process will not only be responsible for relaying IBC packets
+to and from the source chain and Umee, but it will also be responsible for
+listening for these delegation events, and broadcasting the constructed multisig
+transaction to the source chain.
+
+Any new transfers during E<sub>1</sub> will be processed in the same manner during
+E<sub>2</sub>.
+
+## Consequences
+
+### Positive
+
+- Allows for cross-chain delegations in a decentralized manner.
+
+### Negative
+
+- Requires the maintenance of a validator controlled multisig account each epoch.
```

Comment:
I assume this requires the a hot key too

Reply:
Yes, but there is no other choice AFAIK :-/

Reply:
If the multisig is weighted according to stake on Umee then essentially consensus security must be broken in order to compromise the multisig?

Reinforcements may be done since there's a chance that KMS cloud services will get FIPs 140-3 compliant and have ed25519 hardware signers in the cloud soon this year.

Wondering if it's possible to have a security model to prevent functionality other than delegations related transactions on the multisig and returning owed money to the user. Sort of like how proxy contracts work on Ethereum.

Reply:
The multisig is not weighted -- it's a simple n/n multisig. All validators are required to sign per epoch.

Reply:
So let's say we go with the n/n approach - managing this multisig is still quite difficult across chains if you want validator set flexibility? Also with n of n you'd have to be stricter about who can become a validator because misoperations from anyone is opportunity cost and would stop progress.

It also seems dangerous, what happens if you have an n of n multisig and someone's key set up is destroyed? All the user funds are gone then because of one validator?

Reply:
> managing this multisig is still quite difficult across chains if you want validator set flexibility? 

What is being managed cross chain? There is a single validator set multisig per epoch which must exist on the native ATOM chain. Validator sets can change every epoch.

> Also with n of n you'd have to be stricter about who can become a validator because misoperations from anyone is opportunity cost and would stop progress.

Yes, if a single validator misses a tx, the delegation does not happen for that epoch. The chain progress normally though.

> It also seems dangerous, what happens if you have an n of n multisig and someone's key set up is destroyed? All the user funds are gone then because of one validator? 

I think we'd have the validators submit txs first, and then send funds to that account on the native ATOM chain once we received all signatures. Only then do we emit the event.

I get this process is not clean, but I cannot think of any other way to do cross-chain delegation. Do you have alternative ideas?

Reply:
> What is being managed cross chain? There is a single validator set multisig per epoch which must exist on the native ATOM chain. Validator sets can change every epoch.

Let me rephrase - so there's epoch management on the Umee chain and a multisig that's managed on the ATOM chain by the Umee validators, which gets managed by the `x/umee` module if I understand correctly. If we have epochs, then validator set changes are easy. I am not sure how the multisig gets updated though through these epochs for staked assets, but the notion of an epoch is useful for synchronizing events to happen.

> I think we'd have the validators submit txs first, and then send funds to that account on the native ATOM chain once we received all signatures. Only then do we emit the event.

That makes sense. I think this fact is still scary:

> It also seems dangerous, what happens if you have an n of n multisig and someone's key set up is destroyed? All the user funds are gone then because of one validator?

Are weighted multisigs completely off the table? We'd have to add support to the SDK and Hub.